### DISCUSSION TITLE
API changes for v1.0

### DISCUSSION THREAD

**Discussion by lukasf** [ffmpeginteropx/FFmpegInteropX#226] [Ideas]
I have been thinking about doing some API changes for v1.0 of the library. After we give it a 1.x label, it will be hard to remove or rename things. So maybe we should give some thought about how the 1.0 API should look like.

Some of the ideas I had:

## Remove deprecated APIs

This one is obvious. Let's drop those old deprecated APIs. I guess and hope that noone really uses them anymore. There are better alternatives for every deprecated API, so we are not taking away any functionality.

## Update namespace and library name

We have renamed the library from FFmpegInterop to FFmpegInteropX. But we did not change the namespace accordingly. I think that we should take this opportunity to update the namespace, and also rename the lib (dll) to FFmpegInteropX.

### Rename FFmpegInteropMSS

I guess that some people won't like the idea to rename our main class. But I would like to rename that dreaded FFmpegInteropMSS to something actually meaningful and intuitive. The old name does not make much sense. MSS means MediaStreamSource, but this is not a MediaStreamSource. Some people won't even know the term MSS.

So what does FFmpegInteropMSS do?

* It is a source for media files or streams
* As a source, it is used for decoding, not encoding
* It uses FFmpeg for its internal workings
* It allows you to create a MediaStreamSource or MediaPlaybackItem for playback
* But you could also use it to just look at the streams of a file or its metadata

Proposals for a new name: FFmpegSource or FFmpegMediaSource. Or maybe FFmpegDecoder?

### Rename BasicVideoEffect?

The BasicVideoEffect implements IBasicVideoEffect. That's probably where its name comes from. Is there a better name for this? Maybe something like VideoPostprocessingEffect? It's really not that important, but still I've been thinking about this.

### Remove CodecChecker?

I put quite a lot of work into this. But with native HW acceleration in place, this does not really have much purpose anymore. Also, it does not support Multi-GPU setup, it always uses the app's default GPU. We could leave the code in, so we can reactivate it later, if we see use for it.

**Comment by OneFingerCodingWarrior** [comment]
how about rename FFmpegInteropMSS to FFmpegAdapter ?

**Comment by brabebhin** [comment]
I think renaming BasicVideoEffect to something else is a breaking change, for the simple reason that it will still pass compilation but the actual effect will not work, since it's activation is tied to its name, and this will introduce some subtle bugs.

As for renamig FFmpegInteropMSS, FFmpegMediaSource is fine.

Another point to consider: maybe we should go full on headers-only library. Right now we have some classes that are header only and some that are not.

**Comment by lukasf** [comment]
@OneFingerCodingWarrior I have been thinking about "Adapter" as well. But the class does more than just adapt to something. Like enumerate streams and metadata.

@mcosmin222 Thanks for the hint on BasicVideoEffect. We sure do not want to introduce subtle bugs, so we will leave the name untouched.

I am generally open for moving more stuff to header-only. It's just that I am too lazy to do it. :) It is work that does not add any features. So I would personally only do it while doing bigger refactorings in a class. But feel free to move stuff header-only, if you like to do it. Just one thing: This does not change any APIs, so it can be done later. And right now, I would like to stabilize things and head for a 1.0 release. So let's delay this until we have a stable 1.0 online.

**Reply by OneFingerCodingWarrior** [reply]
well, i vote for FFmpegMediaSource

**Comment by lukasf** [comment]
When renaming the main class, the name of the config class should be aligned as well. So we would end up with either FFmpegMediaSource + FFmpegMediaSourceConfig or FFmpegSource + FFmpegSourceConfig or FFmpegDecoder + FFmpegDecoderConfig. The names of the first pair seem a bit long to me. I think I would perfer one of the last two combinations.

Other question: Should we remove the FFmpegVersionInfo class? Now that we manage FFmpeg versions though NuGet packets, I do not see much sense in it anymore. Also, it uses a version string that is not intended to be parsed, according to av_version_info() docs. So I'd say we either remove it completely, or remove everything but "CurrentVersion" (and remove any parsing there).

**Comment by brabebhin** [comment]
I think it is safe to remove FFmpegVersionInfo.

We could name the config class "MediaSourceConfig" in the case of the first pairing. We don't really need to have "FFmpeg" everywhere, we can put that in the namespace.

**Comment by lukasf** [comment]
Agreed, I will deprecate FFmpegVersionInfo then.

Good idea on the naming! So then we will have FFmpegMediaSource and MediaSourceConfig.
